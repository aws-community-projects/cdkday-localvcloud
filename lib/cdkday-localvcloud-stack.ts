import { Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, Table } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { LogGroup } from 'aws-cdk-lib/aws-logs';
import {
  Choice,
  Condition,
  LogLevel,
  Pass,
  StateMachine,
} from 'aws-cdk-lib/aws-stepfunctions';
import { LambdaInvoke } from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';

export class CdkdayLocalvcloudStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const table = new Table(this, 'DrivingAgeTable', {
      partitionKey: { name: 'pk', type: AttributeType.STRING },
      tableName: 'DrivingAge',
    });

    const fn = new NodejsFunction(this, 'CheckAgeFn', {
      environment: { TABLE_NAME: table.tableName },
      entry: 'lib/check-age.ts',
      functionName: 'CheckAgeFn',
    });

    table.grantReadData(fn);

    // const getItem = new DynamoGetItem(this, 'GetByCountry', {
    //   key: {
    //     pk: DynamoAttributeValue.fromString(JsonPath.stringAt('$.Country')),
    //   },
    //   resultPath: '$.Result',
    //   table,
    // });

    const checkAge = new LambdaInvoke(this, 'CheckAge', {
      lambdaFunction: fn,
      resultPath: '$.Result',
    });

    const oldEnough = new Choice(this, 'Can you drive?');

    const canDrive = new Pass(this, 'Can Drive');

    const cannotDrive = new Pass(this, 'Cannot Drive');

    oldEnough.when(
      Condition.numberGreaterThanEqualsJsonPath('$.Age', '$.Result.Payload'),
      canDrive
    );

    oldEnough.otherwise(cannotDrive);

    new StateMachine(this, 'DrivingAgeStateMachine', {
      definition: checkAge.next(oldEnough),
      logs: {
        destination: new LogGroup(this, 'DrivingAgeStateMachineLogs'),
        includeExecutionData: true,
        level: LogLevel.ALL,
      },
      stateMachineName: 'DrivingAgeStateMachine',
    });
  }
}
