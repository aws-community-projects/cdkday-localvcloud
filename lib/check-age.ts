import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

interface InputEvent {
  Country: string;
}

const { AWS_SAM_LOCAL, TABLE_NAME } = process.env;

const docClient = DynamoDBDocumentClient.from(
  new DynamoDBClient({
    endpoint: AWS_SAM_LOCAL ? 'http://ddb:8000' : '',
  })
);

export const handler = async (event: InputEvent): Promise<number> => {
  const { Country } = event;
  const command = new GetCommand({
    Key: { pk: Country },
    TableName: TABLE_NAME,
  });
  const response = await docClient.send(command);
  if (response.Item) {
    return response.Item.Age;
  }
  throw new Error(`Country ${Country} not found!`);
};
