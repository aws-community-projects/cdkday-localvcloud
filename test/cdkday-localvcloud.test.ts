import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { expect, test } from 'vitest';

import * as CdkdayLocalvcloud from '../lib/cdkday-localvcloud-stack';

test('Create Stack', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new CdkdayLocalvcloud.CdkdayLocalvcloudStack(
    app,
    'MyTestStack'
  );
  // THEN
  const template = Template.fromStack(stack);

  expect(template).toMatchSnapshot();
});
