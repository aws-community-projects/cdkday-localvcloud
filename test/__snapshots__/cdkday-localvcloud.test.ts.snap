// Vitest Snapshot v1

exports[`Create Stack 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CheckAgeFnCC0F36E8": {
      "DependsOn": [
        "CheckAgeFnServiceRoleDefaultPolicy398E730F",
        "CheckAgeFnServiceRole6E0E921A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "89d289edbc69fb507e52525fc6a0a1d3a8099461d8e32d4c3c6307b6071c4a0e.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Ref": "DrivingAgeTable8E413A64",
            },
          },
        },
        "FunctionName": "CheckAgeFn",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CheckAgeFnServiceRole6E0E921A",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CheckAgeFnServiceRole6E0E921A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CheckAgeFnServiceRoleDefaultPolicy398E730F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "DrivingAgeTable8E413A64",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CheckAgeFnServiceRoleDefaultPolicy398E730F",
        "Roles": [
          {
            "Ref": "CheckAgeFnServiceRole6E0E921A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DrivingAgeStateMachine90F0E904": {
      "DependsOn": [
        "DrivingAgeStateMachineRoleDefaultPolicy77ABCBD5",
        "DrivingAgeStateMachineRole1F32F5F1",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{\\"StartAt\\":\\"CheckAge\\",\\"States\\":{\\"CheckAge\\":{\\"Next\\":\\"Can you drive?\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"ResultPath\\":\\"$.Result\\",\\"Resource\\":\\"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              {
                "Fn::GetAtt": [
                  "CheckAgeFnCC0F36E8",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"Can you drive?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Age\\",\\"NumericGreaterThanEqualsPath\\":\\"$.Result.Payload\\",\\"Next\\":\\"Can Drive\\"}],\\"Default\\":\\"Cannot Drive\\"},\\"Cannot Drive\\":{\\"Type\\":\\"Pass\\",\\"Result\\":{\\"canDrive\\":false},\\"End\\":true},\\"Can Drive\\":{\\"Type\\":\\"Pass\\",\\"Result\\":{\\"canDrive\\":true},\\"End\\":true}}}",
            ],
          ],
        },
        "LoggingConfiguration": {
          "Destinations": [
            {
              "CloudWatchLogsLogGroup": {
                "LogGroupArn": {
                  "Fn::GetAtt": [
                    "DrivingAgeStateMachineLogs34DDB8F4",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": true,
          "Level": "ALL",
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "DrivingAgeStateMachineRole1F32F5F1",
            "Arn",
          ],
        },
        "StateMachineName": "DrivingAgeStateMachine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "DrivingAgeStateMachineLogs34DDB8F4": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "DrivingAgeStateMachineRole1F32F5F1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "DrivingAgeStateMachineRoleDefaultPolicy77ABCBD5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CheckAgeFnCC0F36E8",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CheckAgeFnCC0F36E8",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DrivingAgeStateMachineRoleDefaultPolicy77ABCBD5",
        "Roles": [
          {
            "Ref": "DrivingAgeStateMachineRole1F32F5F1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DrivingAgeTable8E413A64": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "DrivingAge",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
